[Federation][kubefed] Support configuring dns-provider

shashidharatd opened the pull request:
**What this PR does / why we need it**:
Some environments might need to configure the dns-provider using custom configurations for deploying federation control plane. This PR will facilitate such scenarios. please refer to #40620 

Now we can pass dns provider configuration using `dns-provider-config` flag to `kubefed init`

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #40620

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```
[Federation] Introduced a new flag --dns-provider-config to kubefed to configure dns provider via a config file in local file system.
```

cc @kubernetes/sig-federation-misc @madhusudancs @irfanurrehman @marun

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40528)
<!-- Reviewable:end -->

marun commented:
I think there are some problems with this approach:

 - chicken-or-egg:  ``kubefed init`` will fail if the target namespace already exists.  It should be possible either for ``kubefed`` to use an existing namespace, or to have ``kubefed`` create the resource itself (i.e. accept the path to a local file that it will create a resource from).  I would lean towards the latter, but I think other input is required before moving forward.
 - given that dns configuration contains potentially sensitive credentials, why is a configmap preferable to a secret?

Same concern as https://github.com/kubernetes/kubernetes/pull/40516 regarding creating a PR without an issue.  We'll hash out the details here now that you've posted, but I think it's preferable to have an issue as a point of coordination before work begins.  This minimum of process can help to ensure that effort (not just implementation, but also review/documentation/testing/etc) can be prioritized and that contributors don't step on each other.

marun commented:
This appears to be a partial solution to https://github.com/kubernetes/kubernetes/issues/39870

Commit:
Enable configuring dns-provider from kubefed

shashidharatd commented:
@marun, I have addressed the comments raised by you. Now i am creating a Secret instead of ConfigMap and also taking dns-provider-config file to kubefed init instead of ConfigMap. PTAL

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -121,6 +124,7 @@ func NewCmdInit(cmdOut io.Writer, config util.AdminConfig) *cobra.Command {
 	cmd.Flags().String("dns-zone-name", "", "DNS suffix for this federation. Federated Service DNS names are published with this suffix.")
 	cmd.Flags().String("image", defaultImage, "Image to use for federation API server and controller manager binaries.")
 	cmd.Flags().String("dns-provider", "google-clouddns", "Dns provider to be used for this deployment.")
+	cmd.Flags().String("dns-provider-config", "", "Config file path on local file system for configuring DNS provider.")
```
(No action required) Do you know why the ``AddFlags()`` approach (e.g. https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L94) wasn't used here?  It seems cleaner to be able to use a ``FlagSet`` to define a flag and its assignment in a single line.

cc: @madhusudancs

shashidharatd replied:
Agree. its less prone to human errors. @madhusudancs, should i implement it in this pr?

marun replied:
I think it should be separated separately if at all - it's orthogonal to the problem being solved here.

shashidharatd replied:
Agree.

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -583,9 +587,51 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 		},
 	}
 
+	if dnsProviderConfig != "" {
```
Having to validate this change by exercising the whole of the controller manager init is more work than necessary.  Consider factoring these additions into separate function(s) to simplify unit testing, e.g.

```golang
func addDNSProviderConfig(dnsProviderConfig string, dep *extensions.Deployment) {
  if dnsProviderConfig != "" {
    ...
  }
}
```

shashidharatd replied:
imho, its ok to segregate the dnsProviderConfig to a separate function, but still we need to do similar mechanism to test out this option (using fake http). @marun, can you suggest any alternative you have in your mind for testing just this function `addDNSProviderConfig`

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -583,9 +587,51 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+				dnsProviderSecretKey: dnsProviderConfigBytes,
+			},
+		}
+		secret, err = clientset.Core().Secrets(namespace).Create(secret)
```
(No action required) If an error occurs after the secret is created, the secret would linger.  I think this is a general problem with how ``kubefed`` operates.  Not saying it should be addressed in this PR, but I think it may make sense for ``kubefed`` to be able to clean up in the event of error (i.e. make ``init`` et al all-or-nothing).

shashidharatd replied:
The current mechanism to cleanup is to do something like delete the federation-system namespace and start over again.

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -583,9 +587,51 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+
+		// Create a volume from dns-provider secret
+		volume := api.Volume{
+			Name: "config-volume",
```
Don't repeat yourself (DRY): Define the volume name with a constant

shashidharatd replied:
Agree

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -583,9 +587,51 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+		// Mount dns-provider secret volume to controller-manager container
+		volumeMount := api.VolumeMount{
+			Name:      "config-volume",
+			MountPath: "/tmp/dns-provider-config",
```
DRY: Define the mount path with a constant

shashidharatd replied:
Agree

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -583,9 +587,51 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 	if dryRun {
 		return dep, nil
 	}
+
```
Unrelated change - please revert

shashidharatd replied:
ok

marun commented:
As per #40620 a json file didn't work when passed via ``--dns-provider-config`` when the provider was google dns.  I'm assuming there is a process to convert from json to expected conf format but a cursory search didn't turn up with anything.  I think that in addition to the proposed change, docs for each provider will be necessary to ensure it can be used.

shashidharatd commented:
@marun, this pr takes a general approach to configure dns provider. Each of the dns provider may have separate format (type of file, or which fields need to be present etc.). That should be explained in kubefed tutorial doc. @madhusudancs, request for your comment here.

shashidharatd commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 590 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r1-590:-Kc6UcNd9HQIBkiqh0Yz:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5aa73fa386a41e3554b2f682f905ef8da182d4e7/federation/pkg/kubefed/init/init.go#L590)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

imho, its ok to segregate the dnsProviderConfig to a separate function, but still we need to do similar mechanism to test out this option (using fake http). @marun, can you suggest any alternative you have in your mind for testing just this function `addDNSProviderConfig`
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 612 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r1-612:-Kc6UmeMmVIR4w4888AA:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5aa73fa386a41e3554b2f682f905ef8da182d4e7/federation/pkg/kubefed/init/init.go#L612)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Agree
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 624 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r1-624:-Kc6UoC5m2h5xPwkCMBY:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5aa73fa386a41e3554b2f682f905ef8da182d4e7/federation/pkg/kubefed/init/init.go#L624)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Agree
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 634 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r1-634:-Kc6UpVyyse8thTHie1E:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5aa73fa386a41e3554b2f682f905ef8da182d4e7/federation/pkg/kubefed/init/init.go#L634)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528)*
<!-- Sent from Reviewable.io -->

marun commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 797 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcEo75sy_Cwc1VpKwPb:-KcEo75sy_Cwc1VpKwPc:bkq3ggl) ([raw file](https://github.com/kubernetes/kubernetes/blob/759a704b27f93b5e74fd0ac5ebd67607ae42e21d/federation/pkg/kubefed/init/init.go#L797)):*
> ```Go
> 
> func createDNSProviderConfigSpec(namespace, dnsProviderConfig string) (*api.Secret, error) {
> 	dnsProviderConfigBytes, err := ioutil.ReadFile(dnsProviderConfig)
> ```

I think it would be preferable to read the file before any cluster state is modified to minimize the cost of error recovery.

---

*[federation/pkg/kubefed/init/init_test.go, line 1209 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcEpLUES_KFJkCa5xYE:-KcEpLUES_KFJkCa5xYF:b-yrwjbw) ([raw file](https://github.com/kubernetes/kubernetes/blob/759a704b27f93b5e74fd0ac5ebd67607ae42e21d/federation/pkg/kubefed/init/init_test.go#L1209)):*
> ```Go
> 
> // TODO: Reuse the function addDNSProviderConfig once that function is converted to use versioned objects.
> func addDNSProviderConfigTest(dep *v1beta1.Deployment) *v1beta1.Deployment {
> ```

This approach isn't testing the code, but is still adding maintenance overhead.  Is there another option that doesn't involve a complete test rewrite?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
PR Rebased, and have handled comments. PTAL

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 797 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcEo75sy_Cwc1VpKwPb:-KcGVU9etuY8Pgas3qCx:bis0m37) ([raw file](https://github.com/kubernetes/kubernetes/blob/759a704b27f93b5e74fd0ac5ebd67607ae42e21d/federation/pkg/kubefed/init/init.go#L797)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I think it would be preferable to read the file before any cluster state is modified to minimize the cost of error recovery.
</blockquote></details>

Agree. Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1209 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcEpLUES_KFJkCa5xYE:-KcGVd4X4dSu74Ix0AoZ:b-jje9pt) ([raw file](https://github.com/kubernetes/kubernetes/blob/759a704b27f93b5e74fd0ac5ebd67607ae42e21d/federation/pkg/kubefed/init/init_test.go#L1209)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

This approach isn't testing the code, but is still adding maintenance overhead.  Is there another option that doesn't involve a complete test rewrite?
</blockquote></details>

This function addDNSProviderConfigTest is a duplicate of addDNSProviderConfig which is in init.go and we need to remove this function in future replacing it by addDNSProviderConfig.  I have put a TODO comment, for future reference.
Currently the problem is hostFactory.ClientSet (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L88) returns internal clientset and in test code we are exercising versioned RestClient. So had to duplicate this function to return versioned deployment object.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KcGXaJp-KItbryuODwB:batzcbx)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@marun, request to PTAL at fixes and the responses.

marun commented:
Minor nit about the error message, otherwise lgtm

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/kubefed/init/init.go, line 201 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcTB5VZhs923CUcvwXv:-KcTB5VZhs923CUcvwXw:b8cdkgk) ([raw file](https://github.com/kubernetes/kubernetes/blob/761b63afacfe69988dfc6de6268893395d9698b3/federation/pkg/kubefed/init/init.go#L201)):*
> ```Go
> 		dnsProviderConfigBytes, err = ioutil.ReadFile(dnsProviderConfig)
> 		if err != nil {
> 			return err
> ```

Please provide some context for this error (e.g. ``Error reading file provided to --dns-provider-config: %v``).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KcTBhmmfga0zMb56un1:b-vzo65y)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
handled nit. @marun PTAL
@madhusudancs, request for review

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/kubefed/init/init.go, line 201 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KcTB5VZhs923CUcvwXv:-KcTu0b25YtB_0303IrZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/761b63afacfe69988dfc6de6268893395d9698b3/federation/pkg/kubefed/init/init.go#L201)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Please provide some context for this error (e.g. ``Error reading file provided to --dns-provider-config: %v``).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KcTu2Mfqf17qhMuQxgo:bvkxnia)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, this pr is rebased and have handled all comments from @marun, request to have review from you, as this has been open since 15 days and also enables me to write documentation for using CoreDNS DNS provider in federation, which many people are interested to know.

marun commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528)*
<!-- Sent from Reviewable.io -->

lukasredynk commented:
One thing: maybe it'll be a good idea to add check if there's specified dns provider config if coredns is selected and fail, if not.

shashidharatd commented:
@lukasredynk, why do you think this pr about adding --dns-provider-config, is only required for CoreDNS?
It can be used by all other dns providers to customize configuration

madhusudancs commented:
@shashidharatd a few minor nits, but mostly LGTM. Please let me know once you fix them, I will apply the label.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[federation/pkg/kubefed/init/init.go, line 127 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r1-127:-KdXChp5iYxL6KM-ovv1:bg6domx) ([raw file](https://github.com/kubernetes/kubernetes/blob/5aa73fa386a41e3554b2f682f905ef8da182d4e7/federation/pkg/kubefed/init/init.go#L127)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Agree.
</blockquote></details>

@marun has taken care of this now. So considering it resolved.

---

*[federation/pkg/kubefed/init/init.go, line 92 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXBJ3MWWWxyqvAZt_3:-KdXBJ3MWWWxyqvAZt_4:be86ksy) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L92)):*
> ```Go
> 	dnsProviderSecretKey       = "federation-dns-provider.conf"
> 	dnsProviderConfigVolume    = "config-volume"
> 	dnsProviderConfigMountPath = "/tmp/dns-provider-config"
> ```

I would choose a path other than `/tmp`. It is common to put configs in /etc. Can we put this somewhere in /etc as well? `/etc/federation/controller-manager/` or something?

---

*[federation/pkg/kubefed/init/init.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXCr280lhzz5163t1-:-KdXCr280lhzz5163t10:b-8y4u8a) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L233)):*
> ```Go
> 		_, err = ioutil.ReadFile(i.options.dnsProviderConfig)
> 		if err != nil {
> 			return fmt.Errorf("Error reading file provided to --dns-provider-config flag, err: %v", err)
> ```

s/Error reading file/error reading file/

Why do this check here? As in why read the contents? Are you trying to check if the file exists and is a regular file? If so, please use os.Stat().

I think it should also be Ok to read the contents into the `initFederation.dnsProviderConfigBytes` here and use that below. But I am not comfortable reading the on-disk file twice.

---

*[federation/pkg/kubefed/init/init.go, line 257 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXDXxuV6GmHiTVklmw:-KdXDXxuV6GmHiTVklmx:b-otilcs) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L257)):*
> ```Go
> 	var dnsProviderConfigBytes []byte
> 	if i.options.dnsProviderConfig != "" {
> 		dnsProviderConfigBytes, err = ioutil.ReadFile(i.options.dnsProviderConfig)
> ```

See comment above.

---

*[federation/pkg/kubefed/init/init.go, line 829 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXE9Zab8Bu-jZ7ToN1:-KdXE9Zab8Bu-jZ7ToN2:b-6lpjyn) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L829)):*
> ```Go
> 			}
> 		}
> ```

Can we pull this block into its own function and call it from the top-level before calling createControllerManger and pass the secret name to this function?

---

*[federation/pkg/kubefed/init/init_test.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXLV6jw_iPCpnqB-B-:-KdXLV6kTgQKb6adq8V7:b-a8c7c6) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L233)):*
> ```Go
> 		if tc.dnsProviderConfig != "" {
> 			// Create an empty file and provide as an option
> 			_, err = os.Create(tc.dnsProviderConfig)
> ```

Can you just create a tmp file using https://golang.org/pkg/io/ioutil/#example_TempFile?

---

*[federation/pkg/kubefed/init/init_test.go, line 282 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXLqsw4ISf6CEzbVYY:-KdXLqsxdvEJo14_xEZj:b-fx8aqu) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L282)):*
> ```Go
> 		// Cleanup
> 		if tc.dnsProviderConfig != "" {
> 			err = os.Remove(tc.dnsProviderConfig)
> ```

This should be a defer statement immediately after temp create above.

---

*[federation/pkg/kubefed/init/init_test.go, line 672 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXEfYuy5vd0VnDYAiA:-KdXEfYuy5vd0VnDYAiB:b7y7pgb) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L672)):*
> ```Go
> 	}
> 
> 	cmDnsProviderSecret := v1.Secret{
> ```

`cmDNSProviderSecret`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KdXM-Xia_-gVzp1shVi:b2qsl1r)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Handled comments. @madhusudancs PTAL

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 92 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXBJ3MWWWxyqvAZt_3:-KdZQhMgOSxQUCyQS-gw:bfwll2v) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L92)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would choose a path other than `/tmp`. It is common to put configs in /etc. Can we put this somewhere in /etc as well? `/etc/federation/controller-manager/` or something?
</blockquote></details>

Done. used /etc/federation/dns-provider now.

---

*[federation/pkg/kubefed/init/init.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXCr280lhzz5163t1-:-KdZQtV3f-ysXY_3fBGZ:b-iq44r9) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L233)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Error reading file/error reading file/

Why do this check here? As in why read the contents? Are you trying to check if the file exists and is a regular file? If so, please use os.Stat().

I think it should also be Ok to read the contents into the `initFederation.dnsProviderConfigBytes` here and use that below. But I am not comfortable reading the on-disk file twice.
</blockquote></details>

Done. using os.Stat to check if file exists.

---

*[federation/pkg/kubefed/init/init.go, line 257 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXDXxuV6GmHiTVklmw:-KdZROlcpd_wGrmDitiI:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L257)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

See comment above.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 829 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXE9Zab8Bu-jZ7ToN1:-KdZR5IocYkfuOWzjtia:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L829)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can we pull this block into its own function and call it from the top-level before calling createControllerManger and pass the secret name to this function?
</blockquote></details>

Done

---

*[federation/pkg/kubefed/init/init_test.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXLV6jw_iPCpnqB-B-:-KdZR7w8eU_gQfnifNX1:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L233)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can you just create a tmp file using https://golang.org/pkg/io/ioutil/#example_TempFile?
</blockquote></details>

Done

---

*[federation/pkg/kubefed/init/init_test.go, line 282 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXLqsw4ISf6CEzbVYY:-KdZRDtEMYlNIXLO5Icr:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L282)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

This should be a defer statement immediately after temp create above.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 672 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXEfYuy5vd0VnDYAiA:-KdZRGR0E4j2P6jhNMb6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init_test.go#L672)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

`cmDNSProviderSecret`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KdZRHkEWP-WlVCJG5oM:bx4vy70)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@k8s-bot cvm gke e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot verify test this

shashidharatd commented:
@k8s-bot verify test this

madhusudancs commented:
@shashidharatd very minor nit. Let me know once you fix this, I will apply the label.

---

Reviewed 1 of 2 files at r5.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdXCr280lhzz5163t1-:-KdZpX4WcL5-Q7scNNYP:b-ucfknh) ([raw file](https://github.com/kubernetes/kubernetes/blob/acb94fdc5590713345a8a2b78403d6c711e90001/federation/pkg/kubefed/init/init.go#L233)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Done. using os.Stat to check if file exists.
</blockquote></details>

Thanks!

---

*[federation/pkg/kubefed/init/init.go, line 330 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KdZppqbckN3GfCw7mEK:-KdZppqbckN3GfCw7mEL:b-xp7zss) ([raw file](https://github.com/kubernetes/kubernetes/blob/660ebbb269a9e2b3f9173da90d5436bc22147a1b/federation/pkg/kubefed/init/init.go#L330)):*
> ```Go
> 	// 7c. Create a dns-provider config secret
> 	var dnsProviderSecret *api.Secret
> 	if !i.options.dryRun && dnsProviderConfigBytes != nil {
> ```

Can you please move dryRun check inside the function for consistency?

You can also move `if dnsProviderConfigBytes != nil {}` inside the function as the very first line and return immediately as soon as this is nil. It makes this function look somewhat cleaner.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-KdZqu7M2z5LyKtS5Sle:b-kdb2os)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@k8s-bot verify test this

shashidharatd commented:
@madhusudancs, comment handled. PTAL

madhusudancs commented:
/lgtm
/approve

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 985 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-Kdh9ZgaXDcasK5oCJ_1:-Kdh9ZgaXDcasK5oCJ_2:b99qwq6) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3adaeafe0c0321937bc57902fa66c01953b1aed/federation/pkg/kubefed/init/init.go#L985)):*
> ```Go
> 
> 	var secret *api.Secret
> 	var err error
> ```

It's ok for now, but it is not necessary to have this `err` variable declared here right?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-:-Kdh9m5dPhUvJ635ImD-:b6yv7yi)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@k8s-bot unit test this

marun commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -236,6 +248,14 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 	cmName := fmt.Sprintf("%s-controller-manager", i.commonOptions.Name)
 	cmKubeconfigName := fmt.Sprintf("%s-kubeconfig", cmName)
 
+	var dnsProviderConfigBytes []byte
```
(no action required) Why not read the file in ``Complete`` rather than stat'ing there and reading here?

madhusudancs commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 251 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528#-KbQhuK5DfQR9g5aICZq-r6-251:-KdqvRzxMZNPgx7Wg-wt:b-x2hmc2) ([raw file](https://github.com/kubernetes/kubernetes/blob/d3adaeafe0c0321937bc57902fa66c01953b1aed/federation/pkg/kubefed/init/init.go#L251)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(no action required) Why not read the file in ``Complete`` rather than stat'ing there and reading here?
</blockquote></details>

That was already suggested here - https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40528#-KdXCr280lhzz5163t1- But this is reasonable too.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40528)*
<!-- Sent from Reviewable.io -->
